### ISSUE TITLE
fork / execvp / arrays

### ISSUE THREAD

**Issue by wookietreiber** [scala-native/scala-native#1104]
**TL;DR:** For a shell, I want to read a command from `stdin` and start the process. Is there a way to convert a `Seq[String]` to a native C array like the one `execvp` needs?

---

Today I was starting to experiment with writing an OS shell in scala-native. I got the basics working and got stuck when trying to start processes, which is what you need for a shell.

What I did is based on [this tutorial](https://brennan.io/2015/01/16/write-a-shell-in-c/), if you want to fast-forward to the part that is relevant to this issue, search for "int lsh_launch".

I searched scala-native and found out that there are no fork or exec functions in it yet, and also the Java/Scala process library can't be used yet. So, I started adding my own little interop stuff:

```scala
@extern
object stuff {
  type pid_t = CInt
  def fork(): pid_t = extern
  def execvp(file: CString, args: Ptr[CString]): CInt = extern
  def wait(wstatus: Ptr[CInt]): pid_t = extern
}
```

This should, I think, represent the C functions, although I'm not 100% sure.

I was testing this and I'm having problems constructing the `args` for `execvp`. I have created a small repository, which you can find [here](https://github.com/wookietreiber/scala-native-fork). I've put in some of   my failed tries and unexpected results in that projects README. I've tried so many times that I don't see anymore what I do wrong. Please help, I don't get it.

**Comment by LeeTibbert** [comment]
@wookietreiber 

Interesting  study piece.  I only had time for a quick look at your code,
I did not have time to experiment.

If I recall correctly (and I think the C tutorial you cited mentions) execvp
expects args to be a NULL terminated array of strings.

Your weird output as number of args varies certainly looks like a 
missing NULL array terminator to me. Yes, I have made that mistake ;-).

I am not certain if c"" will work or if some ugly cast is needed.

I hope this helps.

Lee

**Comment by LeeTibbert** [comment]
@wookietreiber 

TL;DR: I suggest you try adding:  ptr(4) = null

Adding the above line gets fork.scala to compile & link for me.  
It is a close approximation of what is used in the scala-native
test suites, all fault for ugliness & misunderstanding on me, not
that code.

Because I am on a 32-bit system with a highly modified scala-native
environment, I had to hack the original (sleep 5 instead of echo) to
avoid known problems with doing output on advertised as unsupported
Intel 32 bit systems.  I am operating without blade guards outside of
Known Space.  Thus said, the execvp *.out appeared to run successfully
for me.

Note: When running the image from SBT, I saw shrapnel/complaints
from ProcessBuilder.  That is probably either my 32-bit environment
or my bleeding edge scala-native setup.  I just mention this to
qualify the meaning of "runs on my system".

If the above suggestion does not solve your issue, I will try again when I get to my 64-bit 
system (days, not minutes). That system should more closely approximate your environment & code.

**Comment by wookietreiber** [comment]
Thanks to @LeeTibbert suggestions I was able to fix it like this:

```scala
package fork

import scalanative.native._
import scalanative.posix._

@extern
object stuff {
  type pid_t = CInt
  def fork(): pid_t = extern
  def execvp(file: CString, args: Ptr[CString]): CInt = extern
  def wait(wstatus: Ptr[CInt]): pid_t = extern
}

object Main extends App {

  def toCArray(a: Array[String])(implicit Z: Zone): Ptr[CString] = {
    val ptr: Ptr[CString] = alloc[CString](a.size + 1)
    var i = 0
    while (i < a.size) {
      ptr(i) = toCString(a(i))
      i += 1
    }
    ptr(a.size) = null
    ptr
  }

  stuff.fork() match {
    case 0 =>
      Zone { implicit z =>
        val cmd = Array("cp", "README.md", "foo.md")
        val ptr = toCArray(cmd)

        val ret = stuff.execvp(ptr(0), ptr)

        if (ret == -1)
          stdio.perror(c"fooexecvp")
      }

    case pid if pid < 0 =>
      stdio.perror(c"foofork")

    case pid =>
      val wstatus: Ptr[CInt] = stackalloc[CInt]

      while (stuff.wait(wstatus) != pid) {}
  }
}
```

**Comment by wookietreiber** [comment]
@ maintainers / @densh: Do you think this `toCArray` function is worth putting into `scala/scalanative/native/package.scala`? I suppose, this would be nice to have for other functions requiring a `char**`. If so, I could prepare a pull request for this.

```scala
  def toCArray(a: IndexedSeq[String])(implicit Z: Zone): Ptr[CString] = {
    val ptr: Ptr[CString] = alloc[CString](a.size + 1)
    var i = 0
    while (i < a.size) {
      ptr(i) = toCString(a(i))
      i += 1
    }
    ptr(a.size) = null
    ptr
  }
```
